### ISSUE TITLE
ICE in solidity::frontend::ArrayType::memoryDataSize, Array size does not fit u256

### ISSUE THREAD

**Issue by Changochen** [ethereum/solidity#8286]
## Description
Consider the following POC:
```
contract C {
        struct X { bytes31 [ 3 ] x1 ;
                uint x2 ;
        }
        struct S { uint256 [ ] [ 0.425781 ether ] s1 ;
                uint [ 2 ** 0xFF ] [ 2 ** 0x42 ] s2 ;
                X s3 ;
                mapping ( uint => address payable ) c ;
                uint [ 9 hours ** 16 ] d ;
                string s ;
        }
        function f ( ) public { function ( function ( bytes9 , uint ) external pure returns ( uint ) , uint ) external pure returns ( uint ) [ 3 ] memory s2 ;
                S memory s ;
        }
}
```

Run with `solc --bin -o poc poc.sol` and we got 
```
Internal compiler error during compilation:
/home/yongheng/solidity_clean/libsolidity/ast/Types.cpp(1855): Throw in function virtual solidity::u256 solidity::frontend::ArrayType::memoryDataSize() const
Dynamic exception type: boost::wrapexcept<solidity::langutil::InternalCompilerError>
std::exception::what: Array size does not fit u256.
[solidity::util::tag_comment*] = Array size does not fit u256.
```
## Environment

- Compiler version: Version: 0.6.3-develop.2020.2.10+commit.d033c2f7.Linux.clang
- Target EVM version (as per compiler settings):
- Framework/IDE (e.g. Truffle or Remix):
- EVM execution environment / backend / blockchain client:
- Operating system: Ubuntu 16

**Comment by bshastry** [comment]
Minimal example

```
contract C {
  struct S { int[2 ** 0xFF] m; }
  function f() public { S memory s; }
}
```

Reproducing via command line

```
$ solc --optimize <test.sol>
libsolidity/ast/Types.cpp(1855): Throw in function virtual solidity::u256 solidity::frontend::ArrayType::memoryDataSize() const
Dynamic exception type: boost::exception_detail::clone_impl<solidity::langutil::InternalCompilerError>
std::exception::what: Array size does not fit u256.
[solidity::util::tag_comment*] = Array size does not fit u256.
```

**Comment by bshastry** [comment]
Thank you for reporting @Changochen .

On second thought, I'm not sure the underlying problem is serious. Perhaps, we shouldn't throw an exception but report an error?

**Comment by Changochen** [comment]
Hi @bshastry . Yeah, I also doubt whether this problem is serious. As one solidity developer said in one issue that "ICE is always a compiler bug", I decided to report it so that you can be at least aware of it.

**Comment by axic** [comment]
We (used to) have checks for too large initialisers. This probably is just a missing case and should definitely be reported properly as an error.

**Comment by chriseth** [comment]
Maybe solved by #8291

**Comment by mijovic** [comment]
In ReferenceResolver.cpp line 275 in mobileType() call we check that number of elements for array fits u256 which in minimal example posted passes check because
57896044618658097711785492504343953926634992332820282019728792003956564819968 < 
115792089237316195423570985008687907853269984665640564039457584007913129639935 

Actual size of array can't fit u256. So it needs to be checked with additional check, or decrease maximum number of elements...
